servicetalk-grpc-internal/src/main/java/io/servicetalk/grpc/internal/GrpcStatusUtils.java

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright © 2024 Apple Inc. and the ServiceTalk project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/*
+ * Copyright 2024 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.servicetalk.grpc.internal;
+
+import io.servicetalk.http.api.HttpHeaders;
+
+import java.nio.ByteBuffer;
+import java.nio.charset.StandardCharsets;
+import javax.annotation.Nullable;
+
+import static io.servicetalk.buffer.api.CharSequences.newAsciiString;
+
+/**
+ * Provides utilities around percent-encoding and decoding the GRPC status message.
+ * <p>
+ * Note that much of the actual encoding and decoding logic is borrowed from the {@code io.grpc.Status} class,
+ * specifically the {@code io.grpc.Status#StatusMessageMarshaller}.
+ */
+public final class GrpcStatusUtils {
+
+    public static final CharSequence GRPC_STATUS_MESSAGE = newAsciiString("grpc-message");
+
+    private static final byte[] HEX =
+            {'0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'A', 'B', 'C', 'D', 'E', 'F'};
+
+    private GrpcStatusUtils() {
+        // singleton
+    }
+
+    /**
+     * Sets and potentially encodes the given status message.
+     *
+     * @param headers the headers on where to set the message.
+     * @param message the message to set.
+     */
+    public static void setStatusMessage(final HttpHeaders headers, final CharSequence message) {
+        final byte[] messageBytes = message.toString().getBytes(StandardCharsets.UTF_8);
+        for (int i = 0; i < messageBytes.length; i++) {
+            final byte b = messageBytes[i];
+            // If there are only non escaping characters, skip the slow path.
+            if (isEscapingChar(b)) {
+                headers.set(GRPC_STATUS_MESSAGE, encodeMessage(messageBytes, i));
+                return;
+            }
+        }
+        headers.set(GRPC_STATUS_MESSAGE, message);
```

Comment:
This still doesn't look quite right: we check each byte after the utf8 encoding but it could be possible that the byte representation in utf8 contains no escapable chars even if `message` itself is not an ASCII compatible string.
```suggestion
            for (int i = 0; i < message.length(); i++) {
            char c = message.charAt(i);
            // If there are only ASCII non-escaping characters we can skip the slow path.
            if (c > 127 || isEscapingChar((byte) c)) {
                // We have non ASCII compatible characters. We must encode the string as UTF8 and then
                // possibly percent encode.
                final byte[] messageBytes = message.toString().getBytes(StandardCharsets.UTF_8);
                // next we need to see if we have to percent encode the bytes form.
                for (; i < messageBytes.length; i++) {
                    if (isEscapingChar(messageBytes[i])) {
                        // super slow path: we also have to percent encode.
                        headers.set(GRPC_STATUS_MESSAGE, encodeMessage(messageBytes, i));
                        return;
                    }
                }
                // The utf8 encoded form doesn't require percent encoding. Use the ISO_8859_1 charset to preserve
                // the byte representation of the string.
                headers.set(GRPC_STATUS_MESSAGE, new String(messageBytes, StandardCharsets.ISO_8859_1));
                return;
            }
        }
        headers.set(GRPC_STATUS_MESSAGE, message);
```